api/v1beta1/webhook.go

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2021 kloeckner.i GmbH
+ * Copyright 2023 DB-Operator Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1beta1
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+
+	"k8s.io/utils/strings/slices"
+)
+
+var (
+	helpers       []string = []string{"Protocol", "Hostname", "Port", "Password", "Username", "Password", "Database"}
+	functions     []string = []string{"Secret", "ConfigMap", "Query"}
+	userFunctions []string = []string{"ConfigMap"}
+)
+
+// Make sure that credentials.templates are correct
+// ConfigMaps templating is not allowed for DbUser, that's why we have
+//
+//	the second argument: cmAllowed. It should be set to false when
+//	validation is called by dbuser_webhook.
+func ValidateTemplates(templates Templates, cmAllowed bool) error {
+	for _, template := range templates {
+		allowedFunctions := functions
+		if !cmAllowed {
+			allowedFunctions = userFunctions
+			if !template.Secret {
+				err := errors.New("ConfigMap templating is not allowed for that kind. Please set .secret to true")
```

Comment:
don't understand this part.

When dbuser webhook called, we not allowed to use CMs in not secrets?

The user is not able to change cmAllowed, so for all other cases, this if condition will not be used, or?

Reply:
Templates will end up in Secrets or ConfigMaps, but user doesn't own a cm, so we need to access the data from db configmap. But since it doesn't belong to users they shouldn't be allowed to edit them. If you set 
```yaml
- tempalte: {{ .Hostname }}
  name: HOSTNAME
  secret: false
```

Template should be written to a configmap, so we're not letting set `secret: false` for `DbUser objects.
This condition atm might look like 
```
if kind == dbUser {
  if secret.false {
    return error
  }
}
```

But actually, the list of allowed functions should be the same, I'll fix it

Reply:
Now it should look better. AllowedFunctions should be the same for any caller, but it `cmAllowed` is set to true, user can set `spec.credentials.templates[].secret` to false and write to the configmap, and when `cmAllowed` is false, user must set `secret` to true